File: src/parser.ts
export interface AST {
  type: string;
  body: any[];
};

export function parser(tokens): AST {
  const body = tokens.reduce((accum, token) => {
    if (token.type === 'word') {
      switch (token.value) {
        case 'Paper':
        case 'Pen':
        case 'Line':
          accum.push({
            type: 'CallExpression',
            name: token.value,
            arguments: [],
          });
          return accum;
        default:
          throw new Error(`${token.value} is not supported token value.`);
      }
    } else if (token.type === 'number') {
      const lastExpression = accum[accum.length - 1];
      lastExpression.arguments.push({
        type: "NumberLiteral",
        value: token.value,
      });
      return accum;
    } else {
      throw new Error(`${token.type} is not supported token type.`);
    }
  }, []);

  return {
    type: 'Drawing',
    body: body
  };
}

File: ruby/lib/sbn_compiler_rb/parser.rb
module SbnCompilerRb
  class Parser
    def parse(tokens)
      body = tokens.reduce([]) { |memo, token|
        if token.type == "word"
          memo << Expression.new("CallExpression", token.value, [])
        elsif token.type == "number"
          last_expression = memo.reverse_each.detect { |exp| exp.type == "CallExpression" }
          last_expression.arguments << Argument.new("NumberLiteral", token.value)
        end
        memo
      }
      AST.new("Drawing", body)
    end
  end
end
